pkg/webhook/webhook.go

```
@@ -0,0 +1,213 @@
+	svr, err := newWebhookServer(mgr)
+	if err != nil {
+		log.Printf("Failed to create webhook server: %v", err)
+		return err
+	}
+
+	wh, err := newValidatingWebhook(mgr)
+	if err != nil {
+		log.Printf("Failed to create validating webhook: %v", err)
+		return err
+	}
+
+	svr.Register(wh)
+	return nil
+}
+
+func newValidatingWebhook(mgr manager.Manager) (*admission.Webhook, error) {
+	return builder.NewWebhookBuilder().
+		Validating().
+		Operations(admissionregistrationv1beta1.Create).
+		ForType(&pravegav1alpha1.PravegaCluster{}).
+		Handlers(&pravegaWebhookHandler{}).
+		WithManager(mgr).
+		Build()
+}
+
+func newWebhookServer(mgr manager.Manager) (*webhook.Server, error) {
+	return webhook.NewServer("admission-webhook-server", mgr, webhook.ServerOptions{
+		CertDir: CertDir,
+		BootstrapOptions: &webhook.BootstrapOptions{
+			Secret: &types.NamespacedName{
+				Namespace: os.Getenv("WATCH_NAMESPACE"),
+				Name:      os.Getenv("SECRET_NAME"),
+			},
+
+			// TODO: garbage collect webhook service
+			Service: &webhook.Service{
+				Namespace: os.Getenv("WATCH_NAMESPACE"),
+				Name:      "admission-webhook-server-service",
+				Selectors: map[string]string{
+					"component": "operator",
+				},
+			},
+		},
+	})
+}
+
+type pravegaWebhookHandler struct {
+	client  client.Client
+	scheme  *runtime.Scheme
+	decoder admissiontypes.Decoder
+}
+
+var _ admission.Handler = &pravegaWebhookHandler{}
+
+// Webhook server will call this func when request comes in
+func (pwh *pravegaWebhookHandler) Handle(ctx context.Context, req admissiontypes.Request) admissiontypes.Response {
+	log.Printf("Webhook is handling incoming requests")
+	pravega := &pravegav1alpha1.PravegaCluster{}
+
+	if err := pwh.decoder.Decode(req, pravega); err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+	copy := pravega.DeepCopy()
+
+	err := pwh.validatePravegaManifest(ctx, copy)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusInternalServerError, err)
```

Not sure we should return a 500 Error if validation fails. E.g. if the user wants to use a non-supported Pravega version, the server should return a 400 BadRequest response. However, if there's an unexpected error during the validation, 500 Error is fine. 

Maybe we could make the `validatePravegaManifest` func to return `(bool, error)`, the `bool` value would indicate whether or not the request is valid, if `false` the server will return 400 BadRequest.

---

Right! Make sense. Thanks for the comment.